Add DI, remove static code and many code duplications, add DI usage e…

yotamoron opened the pull request:
…xample

lc-nyovchev commented on src/main/java/com/microtripit/mandrillapp/lutung/Main.java:
```diff
@@ -0,0 +1,30 @@
+ */
+public class Main {
+    public static void main(String[] args) throws IOException, MandrillApiError {
+        final ClassPathXmlApplicationContext context = new ClassPathXmlApplicationContext("classpath:applicationContext-LutungMandrill-all.xml");
```
I kind of disagree adding a Main class to the repo to show how to "wire" it all up together. I think if you want to have an example, you can better have this as a part of the README.

yotamoron replied:
Sure, makes sense. I'll move it.

lc-nyovchev commented on src/main/java/com/microtripit/mandrillapp/lutung/MandrillApi.java:
```diff
@@ -37,36 +26,30 @@
-		inbound = new MandrillInboundApi(key);
-		exports = new MandrillExportsApi(key);
-		ips = new MandrillIpsApi(key);
+		final MandrillService mandrillService = new MandrillService(rootUrl, new MandrillRequestDispatcher(), key);
```
to preserve the logic from the above check, maybe we can throw a NPE or invalidsomething when the rootUril is null in the constructor of the service?

yotamoron replied:
The service constructor has a Preconditions.checkNotNull() on all its arguments - it will throw an exception is rootUrl is null.

lc-nyovchev replied:
True, but it is not evident by just looking at the above code without knowing the details of the service. But all in all, just a minor thing for consistency. I am very happy with the Spring-y direction you try to steer the project into.

lc-nyovchev commented on src/main/java/com/microtripit/mandrillapp/lutung/controller/MandrillExportsApi.java:
```diff
@@ -104,12 +97,10 @@ public MandrillExportJobInfo rejects(final String notifyEmail)
-		return MandrillUtil.query(rootUrl+ "exports/whitelist.json", 
-				params, MandrillExportJobInfo.class);
-		
+		return queryExecutorFactory.create()
```
+1 for removing the static madness

yotamoron replied:
:+1:

lc-nyovchev commented on src/main/resources/applicationContext-LutungMandrill-all.xml:
```diff
@@ -0,0 +1,74 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I'd put this to be a part of the README in a section titled how to setup Mandrill DI for Spring, but wouldn't put it as a part of the repo code.

yotamoron replied:
I removed the dependency on spring, so I think this could be a nice help for someone who would want to use it (it wouldn't affect any code)

lc-nyovchev commented on src/main/java/com/microtripit/mandrillapp/lutung/model/MandrillApiError.java:
```diff
@@ -91,7 +91,7 @@ public final String getMandrillErrorAsJson() {
 	}
 	
-	protected final MandrillApiError withError(final MandrillError error) {
+	public final MandrillApiError withError(final MandrillError error) {
```
+1 of removing the protected madness

lc-nyovchev commented on src/main/java/com/microtripit/mandrillapp/lutung/model/LutungGsonUtils.java:
```diff
@@ -35,7 +35,7 @@ public static final Gson getGson() {
 	}
 	
-	public static final Gson createGson() {
+	private static final Gson createGson() {
```
+1 of changing the visibility to private

lc-nyovchev commented on src/main/java/com/microtripit/mandrillapp/lutung/http/MandrillService.java:
```diff
@@ -0,0 +1,45 @@
+ * @author rschreijer
+ * @since Mar 19, 2013
+ */
+public final class MandrillService {
```
very good that now we have a MandrillService, however making it final is a bit of an overkill imo.

billoneil commented:
We are not going to make this library depend on Guava, Spring core and Spring context. Not everyone uses DI and those who do not everyone uses Spring. What exactly are you trying to solve here? The only issue I see this resolving is the ability to pass in the mandrill root url which was requested once before. If you want the ability to proxy to some other server we can expose a constructor where you pass in a ```HttpClient``` which can be configured to use a proxy. Then you can use your own DI framework to construct MandrillApis that proxy to some test server or the prod server.

lc-nyovchev commented:
Maybe that's correct, it shouldn't actually depend on Spring core/Context it was depending for it for the Main class to run I guess, but the abundance of static methods make it hard for people to change the internals so at least that I see as a good thing. The guava is more or less meh for me.

yotamoron commented:
I fixed most of the stuff and removed the spring dep (it was a bad idea - I agree).
Current code is much more testable, can be used with a DI framework properly, and - most important for us - the user can now inject their own http implementation (most big projects will have their own means of sending and receiving http requests).

lc-nyovchev commented:
I'd also remove the lutung-mandrill-all.xml in the resources folder. I think everybody knows how to wire their stuff, the important thing is removing the statics.

stiebrs commented on src/main/java/com/microtripit/mandrillapp/lutung/controller/MandrillInboundApi.java:
```diff
@@ -33,12 +32,9 @@ public MandrillInboundApi(final String key) {
 	 */
 	 MandrillInboundDomain[] domains() 
```
all the calls are package-private, perhaps should be public?
